engine-plugins/spin-plugin/src/main/java/org/camunda/spin/plugin/impl/feel/integration/CamundaSpinValueMapper.java

```diff
@@ -0,0 +1,144 @@
+import org.camunda.spin.xml.SpinXmlAttribute;
+import org.camunda.spin.xml.SpinXmlElement;
+import org.camunda.spin.xml.SpinXmlNode;
+
+
+public class CamundaSpinValueMapper extends JavaCustomValueMapper {
+
+  @Override
+  public Optional<Val> toValue(Object x, Function<Object, Val> innerValueMapper) {
+    if (x instanceof SpinJsonNode) {
+      SpinJsonNode node = (SpinJsonNode) x;
+      return Optional.of(this.spinJsonToVal(node, innerValueMapper));
+
+    } else if (x instanceof SpinXmlElement) {
+      SpinXmlElement element = (SpinXmlElement) x;
+      return Optional.of(this.spinXmlToVal(element, innerValueMapper));
+
+    } else {
+      return Optional.empty();
+
+    }
+  }
+
+  @Override
+  public Optional<Object> unpackValue(Val value, Function<Val, Object> innerValueMapper) {
+    return Optional.empty();
+  }
+
+  @Override
+  public int priority() {
+    return 30;
+  }
+
+  protected Val spinJsonToVal(SpinJsonNode node, Function<Object, Val> innerValueMapper) {
+    if (node.isObject()) {
+      Map pairs = node.fieldNames().stream()
+        .collect(toMap(field -> field, field -> spinJsonToVal(node.prop(field), innerValueMapper)));
+      return innerValueMapper.apply(pairs);
+
+    } else if (node.isArray()) {
+      List<Val> values = node.elements().stream().map(e -> spinJsonToVal(e, innerValueMapper)).collect(toList());
+      return new ValList(ListHasAsScala(values).asScala().toList());
+
+    } else if (node.isNull()) {
+      return innerValueMapper.apply(null);
+
+    } else {
+      return innerValueMapper.apply(node.value());
+
+    }
+  }
+
+  protected Val spinXmlToVal(SpinXmlElement element, Function<Object, Val> innerValueMapper) {
+    String name = nodeName(element);
+    Val value = spinXmlElementToVal(element, innerValueMapper);
+    Map<String, Object> map = Collections.singletonMap(name, value);
+
+    return innerValueMapper.apply(map);
+  }
+
+  protected Val spinXmlElementToVal(final SpinXmlElement e, Function<Object, Val> innerValueMapper) {
+    Map<String, Object> membersMap = new HashMap<>();
+    String content = e.textContent().trim();
+    if (!content.isEmpty()) {
+      membersMap.put("$content", new ValString(content));
+    }
+
+    Map<String, ValString> attributes = e.attrs().stream().collect(toMap(attr -> spinXmlAttributeToKey(attr), attr -> new ValString(attr.value())));
+    if (!attributes.isEmpty()) {
+      membersMap.putAll(attributes);
+    }
+
+    Map<String, Val> childrenMap = e.childElements().stream()
+     .collect(
+       groupingBy(
+         el -> nodeName(el),
+         mapping(el -> spinXmlElementToVal(el, innerValueMapper), toList())
+                 ))
+     .entrySet().stream()
+       .collect(toMap(entry -> entry.getKey(), entry -> {
+         List<Val> valList = entry.getValue();
+         if (!valList.isEmpty() && valList.size() > 1) {
+           scala.collection.immutable.List<Val> vals = ListHasAsScala(valList).asScala().toList();
```

Comment:
Yes, as I mentioned [here](https://github.com/camunda/camunda-bpm-platform/pull/664#discussion_r381261181), I'll refactor the code, so that the import is not needed.